common/src/main/java/com/wynntils/features/user/ChatMentionFeature.java

```diff
@@ -0,0 +1,124 @@
+
+    @Override
+    protected void onConfigUpdate(ConfigHolder configHolder) {
+        // rebuild pattern incase it has changed
+        pattern = buildPattern();
+    }
+
+    private Pattern buildPattern() {
+        return Pattern.compile(
+                "(?<!\\[)\\b(" + McUtils.mc().getUser().getName()
+                        + (aliases.length() > 0 ? "|" + aliases.replace(",", "|") : "") + ")\\b(?!:|])",
+                Pattern.CASE_INSENSITIVE);
+    }
+
+    @SubscribeEvent()
+    public void onChat(ChatMessageReceivedEvent e) {
+        Component message = e.getMessage();
+
+        Matcher looseMatcher = pattern.matcher(ComponentUtils.getUnformatted(message));
+
+        if (looseMatcher.find()) {
+            if (markMention) {
+                e.setMessage(rewriteComponentWithHighlight(message));
+            }
+            if (dingMention) {
+                McUtils.playSound(SoundEvents.NOTE_BLOCK_PLING.value());
+            }
+        }
+    }
+
+    private Component rewriteComponentWithHighlight(Component comp) {
+        // Create a new component with only the body and style of comp
+        MutableComponent curr = MutableComponent.create(comp.getContents()).withStyle(comp.getStyle());
+        // .getString() is used here as it gives formattingchars when those exist. It is needed for guild messages
+        // because wynn still uses legacy coloring for it.
+        String text = curr.getString();
+
+        // if current component has no text just process its children and return it
+        if (text == "") {
+            for (Component c : comp.getSiblings()) {
+                curr.append(rewriteComponentWithHighlight(c));
+            }
+
+            return curr;
+        }
+
+        // component has text -> check if it has the mention and modify message to highlight it
+        Matcher match = pattern.matcher(text);
+
+        int nextStart = 0;
+
+        List<MutableComponent> comps = new ArrayList();
+
+        // NOTE: Message is modified here!
+        while (match.find()) {
+            // Mention found -> split component and highlight the mention
+            String before = text.substring(nextStart, match.start());
+            String name = text.substring(match.start(), match.end());
+
+            comps.add(Component.literal(before).withStyle(comp.getStyle()));
+            // styling is not used as it breaks guild chat because wynn in their infinite wisdom decided to make
+            // guild chat not use the proper styling
+            comps.add(Component.literal(rewriteColorCode + name));
+
+            nextStart = match.end();
+        }
+
+        MutableComponent afterComp =
+                Component.literal(text.substring(nextStart)).withStyle(comp.getStyle());
+
+        // process any siblings
+        for (Component c : comp.getSiblings()) {
+            afterComp.append(rewriteComponentWithHighlight(c));
+        }
+
+        // throw all of it together
```

Comment:
This also seem very convoluted. Why reversing and stuff?

I think the basic recursive descent here is:
1) check if current node has text that needs to be updated, and
2) process all siblings recursively.

The current text needs to be updated iff a) it has a non-empty text part, and b) it matches the username regexp.

If this is true, you should be able to replace the Component in question with a new Component containing basically Component[text: "", siblings: [prior-to-match,highlighted-match,after-match]]. (And more of course a slightly more complex if you want to keep supporting multiple mentions in the same Component, with after-first-match-and-before-second-match, etc).

And then you can just append all the siblings, recursively processed.

Reply:
I've cleaned up the rewriting function a lot now but I'm not sure of how extracting the rewriting should be done if it needs to be done still. I tried to extract it but either I'm failing to understand how java passes references or I just did something wrong but anyways the code is a lot cleaner now.